server/auth.go

```diff
@@ -45,3 +46,92 @@ func AuthorizedToken(auth oauth2.Authenticator, logger chronograf.Logger, next h
 		return
 	})
 }
+
+// AuthorizedUser extracts the user name and provider from context. If the user and provider can be found on the
+// context, we look up the user by their name and provider. If the user is found, we verify that the user has at
+// at least the role supplied.
+func AuthorizedUser(store chronograf.UsersStore, useAuth bool, role string, logger chronograf.Logger, next http.HandlerFunc) http.HandlerFunc {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		if !useAuth {
+			next(w, r)
+			return
+		}
+
+		log := logger.
+			WithField("component", "role_auth").
+			WithField("remote_addr", r.RemoteAddr).
+			WithField("method", r.Method).
+			WithField("url", r.URL)
+
+		ctx := r.Context()
+
+		username, err := getUsername(ctx)
+		if err != nil {
+			log.Error("Failed to retrieve username from context")
+			Error(w, http.StatusUnauthorized, fmt.Sprintf("User is not authorized"), logger)
+			return
+		}
+		provider, err := getProvider(ctx)
+		if err != nil {
+			log.Error("Failed to retrieve provider from context")
+			Error(w, http.StatusUnauthorized, fmt.Sprintf("User %s is not authorized", username), logger)
+			return
+		}
+
+		u, err := store.Get(ctx, chronograf.UserQuery{Name: &username, Provider: &provider})
+		if err != nil {
+			log.Error("Error to retrieving user")
+			Error(w, http.StatusUnauthorized, fmt.Sprintf("User %s is not authorized", username), logger)
+			return
+		}
+
+		if hasPrivelege(u, role) {
+			next(w, r)
+			return
+		}
+
+		Error(w, http.StatusUnauthorized, fmt.Sprintf("User %s is not authorized", username), logger)
+		return
+
+	})
+}
+
+func hasPrivelege(u *chronograf.User, role string) bool {
+	if u == nil {
+		return false
+	}
+
+	switch role {
+	case ViewerRoleName:
+		for _, r := range u.Roles {
+			switch r.Name {
+			case ViewerRoleName, EditorRoleName, AdminRoleName:
+				return true
+			default:
+				return false
```

Comment:
If a person has more than one role and that role isn't the first one this will return false.
In other words, I think you can just remove the default statements from all these switches in this function.

Reply:
yup. Good call.